### PR TITLE
feat: add notifications filtering by processors

### PR DESCRIPTION
**Pull request by ydayagi**
## Hey, I just made a Pull Request!

Allow processors to distinguish low and high-importance notifications
The first use would be for email processor
Users usually want only high-importance notifications in their emails 
Implementation is based on this suggestion https://github.com/backstage/backstage/pull/24322#issuecomment-2075131200

### CODE REVIEW

File: plugins/notifications-backend/src/service/router.ts
Comment (drodil) [inline plugins/notifications-backend/src/service/router.ts:190]: What happens if the payload does not have severity? The default `normal` is set after processing the options.

**Comment by drodil** [comment]
Missing changeset + API report updates

File: plugins/notifications-backend/src/service/router.ts
Comment (ydayagi) [reply plugins/notifications-backend/src/service/router.ts:190]: in that case the processor is included (not filtered out)

Comment (drodil) [reply plugins/notifications-backend/src/service/router.ts:190]: I mean that the `payload.severity` can be undefined at this phase

**Comment by ydayagi** [comment]
> Missing changeset + API report updates

fixed

File: plugins/notifications-backend/src/service/router.ts
Comment (drodil) [inline plugins/notifications-backend/src/service/router.ts:194]: Why not just `payload.severity ?? 'normal'` like it's done in the send handler?

File: plugins/notifications-backend/src/service/router.ts
Comment (ydayagi) [reply plugins/notifications-backend/src/service/router.ts]: added default

File: plugins/notifications-backend/src/service/router.ts
Comment (ydayagi) [reply plugins/notifications-backend/src/service/router.ts]: fixed

**Comment by ydayagi** [comment]
@freben can u help here? I don't think drodil can merge

**Comment by ydayagi** [comment]
@drodil I made the requested changes. who approves/merges? you?

**Comment by drodil** [comment]
@ydayagi no I cannot do that, it's either @mareklibra @Rugvip or @freben

**Comment by Rugvip** [comment]
@drodil is it good to ship from your PoV?

**Comment by drodil** [comment]
Hm. It's not a major release for starters. Also not sure if the naming convention is understandable as there are no docs. Maybe also there could be other filters included in the first place and not just to get one specific functionality to work; maximumSeverity, topics from the top of my head.

Additionally this change itself does nothing yet and additional changes are required to the actual processors to make this work but that can be done in a separate PR.

**Comment by ydayagi** [comment]
> Hm. It's not a major release for starters. Also not sure if the naming convention is understandable as there are no docs. Maybe also there could be other filters included in the first place and not just to get one specific functionality to work; maximumSeverity, topics from the top of my head.
> 
> Additionally this change itself does nothing yet and additional changes are required to the actual processors to make this work but that can be done in a separate PR.

there is no real use for filters other than minSeverity at the moment.
after this is merged I will create PR for email processor to use minSeverity filter.
I don't see any real life use case for maxSeverity nor for topics

**Comment by drodil** [comment]
@ydayagi you mean **you** don't have a use case for those filters? Backstage is a framework and should be developed as such.

**Comment by ydayagi** [comment]
> @ydayagi you mean **you** don't have a use case for those filters? Backstage is a framework and should be developed as such.

You are right. I meant that maxSeverity to my knowledge/experience might not have a legitimate use case. But you are right. It is better to have it.
Lets be productive here.
What would you like to add in order to have this merged and used by all of backstage users?
Topics and maxSeverity?
There are string fields such as scope and topic.
What about replacing the topic filter with a 'text filter' that searches all descriptive  string fields? I mean title description scope topic.
I think that nowadays people do a more broad text search rather then a specific one. @Rugvip  WDYT too?

**Comment by ydayagi** [comment]
@mareklibra @drodil @Rugvip
I added all filters suggested by drodil plus implementation in email processor

**Comment by mareklibra** [comment]
Honestly, I do not see a use-case for the `maxSeverity` either, but there is just no harm in having it.

Once we have several processors implemented, such filtering will be copy-pasted in each of them.
I would rather see the filtering logic to be shared and implemented once before passing a notification down to a processor.

I will not block merging of this PR on that, let's keep this idea for a future refactoring.

**Comment by mareklibra** [comment]
It LGTM, jsut the change is `minor`, not `major`

**Review by mareklibra — APPROVED** [review]

**Pull request merged**
